src/highdicom/sr/value_types.py

```diff
@@ -55,7 +56,14 @@ def __init__(
         if isinstance(name, Code):
             name = CodedConcept(*name)
         self.ConceptNameCodeSequence = [name]
-        if relationship_type is not None:
+        if relationship_type is None:
+            warnings.warn(
+                'Constructing a content item without specifying a '
+                'relationship type is deprecated behaviour and will be '
+                'removed in a future version of highdicom.',
+                DeprecationWarning
+            )
+        else:
```

Comment:
I am not sure whether we can do this. As far as I understand it, root content items of value type CONTAINER may not have a Relationship Type (e.g., [root content item of root template TID 1500](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#para_85dac8a4-19f9-4283-bd59-d51aa905412d)).

Sections [SR Document Content Tree](dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.html) and [Content Sequence and Relationship Type](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.2.4.html) are not explicit in this regard.

However, since [Relationship Type](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.html#para_f4165794-8736-4c66-84f0-39f64a7fc05d) is an attribute of [Content Sequence](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.html#para_4f32997d-7567-4a1c-b2d1-ce8fc6795299) and the root content item is not contained in a Content Sequence, but the SR Document itself, it cannot have a Relationship Type.

Reply:
We could add this check to all classes derived from `ContentItem` but `ContainerContentItem`.

Reply:
I was very unsure about this, what you are saying certainly makes intuitive sense to me. I was basing this on @dclunie 's comment in #61 in which he said:

> I dont have suggestions about the API as long as when serialized a value for Relationship Type is always supplied, so that invalid objects are not be created ... see http://dicom.nema.org/medical/dicom/current/output/chtml/part03> /sect_C.17.3.2.4.html <http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.2.4.html>
> I.e., at some point it needs to be required, not optional.

Suggesting that relationship type is always required, although he doesn't say anything explicitly about the root. Perhaps he would be able to clarify here whether the root content item is an exception to the rule that a relationship type is always required.

> We could add this check to all classes derived from `ContentItem` but `ContainerContentItem`.

@hackermd on the assumption that the root content item is the one exception to the rule, I think this approach, while a little duplicative, is the right one.

Reply:
> @hackermd on the assumption that the root content item is the one exception to the rule, I think this approach, while a little duplicative, is the right one.

The problem is that when we include the `RelationshipType` attribute into the document, the object may not be compliant with the standard and may fail validation tests. What do you think about making the `relationship_type` parameter required for all classes derived from `ContentItem`, but making it "conditionally required" for `ContainerContentItem`? We could add an `is_root: bool` parameter to `ContainerContentItem` and raise an exception if `is_root == False` and `relationship_type is None`. This would allow the `relationship_type` parameter to remain optional on the `ContentItem` base class.

Reply:
@dclunie SR book to the rescue!

![image](https://user-images.githubusercontent.com/313942/120862613-3c559380-c557-11eb-8948-8fda3f65c496.png)

Reply:
I think it would make a lot of sense if RelationshipType was "1C IFF not Root content item" here: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.html#sect_C.17.3.1.

Reply:
> I think it would make a lot of sense if RelationshipType was "1C IFF not Root content item"

The conceptual hurdle is that the Relationship Type attribute is defined in [Document Relationship Macro Attributes](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.html#para_4f32997d-7567-4a1c-b2d1-ce8fc6795299) under Content Sequence, while all the other attributes are defined the [Document Content Macro Attributes](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.17.3.html#table_C.17-5). The root content item is not contained in a Content Sequence, therefore attribute type 1 is correct in this context I guess.

Reply:
Yes ... the Document Relationship Macro is designed that way for that reason, and no condition is necessary.

Reply:
Thank you @fedorov  and @dclunie  for the clarification!

> What do you think about making the relationship_type parameter required for all classes derived from ContentItem, but making it "conditionally required" for ContainerContentItem? We could add an is_root: bool parameter to ContainerContentItem and raise an exception if is_root == False and relationship_type is None. This would allow the relationship_type parameter to remain optional on the ContentItem base class.

This definitely seems like the right approach and I was thinking about something very similar myself. However, as I'm trying to implement it, I have hit upon a rather nasty problem: changing the relationship type parameters of the subclasses from optional to required attributes in many cases means having to move their positions in the function signatures ahead of other optional attributes (since in python, optional parameters must follow required attributes). This is especially unfortunate given that the `relationship_type` parameters have been placed at the end of the signature by convention. For example, consider the constructor of `NumContentItem`:

```python
class NumContentItem(ContentItem):

    """DICOM SR document content item for value type NUM."""

    def __init__(
        self,
        name: Union[Code, CodedConcept],
        value: Optional[Union[int, float]] = None,
        unit: Optional[Union[Code, CodedConcept]] = None,
        qualifier: Optional[Union[Code, CodedConcept]] = None,
        relationship_type: Optional[Union[str, RelationshipTypeValues]] = None,  # UH OH
    ) -> None:
```

This would be a potentially nasty breaking change (although I doubt that many or even any users are directly interacting with these content item types). I don't see any other good option though, apart from leave it as an optional parameter but have it raise an error if the caller actually leaves it at the default value, which seems very awkward also.

Preference for how to proceed @hackermd ?

Reply:
Hmm...that's indeed a tricky one. Maybe leaving it as is and asserting the presence of the Relationship Type attribute only when the item gets added to Content Sequence? Since all classes derived from `Template` inherit from `ContentSequence`, we could add the check to [ContentSequence.append()](https://github.com/MGHComputationalPathology/highdicom/blob/1a904e58bc3685eef190e626a0052a21831d8168/src/highdicom/sr/value_types.py#L125). It will require a bit more discipline for the nested content items, but I feel this would still be cleanest option at this point (and it would be in agreement with how it is specified in the standard). We could also add a deprecation warning to keep the option of refactoring the content item interface for the 1.0 release. @CPBridge what do you think?